File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/PolarActivityUtils.kt
package com.polar.sdk.impl.utils

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.sdk.api.model.activity.PolarActiveTime
import com.polar.sdk.api.model.activity.PolarActiveTimeData
import com.polar.sdk.api.model.activity.PolarActivitySamplesData
import com.polar.sdk.api.model.activity.PolarActivitySamplesDayData
import com.polar.sdk.api.model.activity.PolarDailySummaryData
import com.polar.sdk.api.model.activity.parsePbActivityInfo
import com.polar.sdk.api.model.activity.parsePbDailySummary
import com.polar.sdk.api.model.activity.polarActiveTimeFromProto
import fi.polar.remote.representation.protobuf.ActivitySamples
import fi.polar.remote.representation.protobuf.DailySummary
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.emptyFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.toList
import protocol.PftpRequest
import protocol.PftpResponse.PbPFtpDirectory
import java.io.ByteArrayOutputStream
import java.time.LocalDate
import java.time.format.DateTimeFormatter

private const val ARABICA_USER_ROOT_FOLDER = "/U/0/"
private const val ACTIVITY_DIRECTORY = "ACT/"
private const val DAILY_SUMMARY_DIRECTORY = "DSUM/"
private const val DAILY_SUMMARY_PROTO = "DSUM.BPB"
private val dateFormatter = DateTimeFormatter.ofPattern("yyyyMMdd")
private const val TAG = "PolarActivityUtils"

enum class CaloriesType {
    ACTIVITY, TRAINING, BMR
}

internal object PolarActivityUtils {

    /**
     * Read step count for given [date].
     */
    suspend fun readStepsFromDayDirectory(client: BlePsFtpClient, date: LocalDate): Int {
        BleLogger.d(TAG, "readStepsFromDayDirectory: $date")
        val activityFileDir = "$ARABICA_USER_ROOT_FOLDER${date.format(dateFormatter)}/$ACTIVITY_DIRECTORY"

        return try {
            val files = listFiles(client, activityFileDir) { entry ->
                entry.matches(Regex("^$activityFileDir/")) ||
                        entry == "ASAMPL" ||
                        entry.contains(".BPB")
            }.toList()

            if (files.isEmpty()) {
                BleLogger.w(TAG, "readStepsFromDayDirectory() could not find files to read for date $date.")
                return 0
            }

            var stepCount = 0
            for (file in files) {
                try {
                    val response = client.request(
                        PftpRequest.PbPFtpOperation.newBuilder()
                            .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                            .setPath(file)
                            .build()
                            .toByteArray()
                    )
                    val proto = ActivitySamples.PbActivitySamples.parseFrom(response.toByteArray())
                    stepCount += proto.stepsSamplesList.sum()
                } catch (error: Throwable) {
                    BleLogger.w(TAG, "readStepsFromDayDirectory() failed for file: $file, error: $error")
                    return 0
                }
            }
            stepCount
        } catch (error: Throwable) {
            BleLogger.w(TAG, "readStepsFromDayDirectory() failed while listing files, error occurred $error.")
            0
        }
    }

    suspend fun readDistanceFromDayDirectory(client: BlePsFtpClient, date: LocalDate): Float {
        BleLogger.d(TAG, "readDistanceFromDayDirectory: $date")
        val dailySummaryFilePath =
            "$ARABICA_USER_ROOT_FOLDER${date.format(dateFormatter)}/$DAILY_SUMMARY_DIRECTORY$DAILY_SUMMARY_PROTO"
        var distance = 0F
        try {
            val response = client.request(
                PftpRequest.PbPFtpOperation.newBuilder()
                    .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                    .setPath(dailySummaryFilePath)
                    .build()
                    .toByteArray())
            distance = DailySummary.PbDailySummary.parseFrom(response.toByteArray()).activityDistance
        } catch (error: Throwable) {
            BleLogger.w(TAG, "readDistanceFromDayDirectory() failed for path: $dailySummaryFilePath, error: $error")
        }
        return distance
    }

    suspend fun readActiveTimeFromDayDirectory(client: BlePsFtpClient, date: LocalDate): PolarActiveTimeData {
        BleLogger.d(TAG, "readActiveTimeFromDayDirectory: $date")
        val dailySummaryFilePath =
            "$ARABICA_USER_ROOT_FOLDER${date.format(dateFormatter)}/$DAILY_SUMMARY_DIRECTORY$DAILY_SUMMARY_PROTO"
        return try {
            val response = client.request(
                PftpRequest.PbPFtpOperation.newBuilder()
                    .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                    .setPath(dailySummaryFilePath)
                    .build()
                    .toByteArray()
            )
            val proto = DailySummary.PbDailySummary.parseFrom(response.toByteArray())
            PolarActiveTimeData(
                date = date,
                timeNonWear = polarActiveTimeFromProto(proto.activityClassTimes.timeNonWear),
                timeSleep = polarActiveTimeFromProto(proto.activityClassTimes.timeSleep),
                timeSedentary = polarActiveTimeFromProto(proto.activityClassTimes.timeSedentary),
                timeLightActivity = polarActiveTimeFromProto(proto.activityClassTimes.timeLightActivity),
                timeContinuousModerateActivity = polarActiveTimeFromProto(proto.activityClassTimes.timeContinuousModerate),
                timeIntermittentModerateActivity = polarActiveTimeFromProto(proto.activityClassTimes.timeIntermittentModerate),
                timeContinuousVigorousActivity = polarActiveTimeFromProto(proto.activityClassTimes.timeContinuousVigorous),
                timeIntermittentVigorousActivity = polarActiveTimeFromProto(proto.activityClassTimes.timeIntermittentVigorous)
            )
        } catch (error: Throwable) {
            BleLogger.w(TAG, "readActiveTimeFromDayDirectory() failed for path: $dailySummaryFilePath, error: $error")
            PolarActiveTimeData(date, PolarActiveTime())
        }
    }

    suspend fun readSpecificCaloriesFromDayDirectory(
        client: BlePsFtpClient,
        date: LocalDate,
        caloriesType: CaloriesType
    ): Int {
        BleLogger.d(TAG, "readSpecificCaloriesFromDayDirectory: $date, type: $caloriesType")
        val dailySummaryFilePath =
            "$ARABICA_USER_ROOT_FOLDER${date.format(dateFormatter)}/$DAILY_SUMMARY_DIRECTORY$DAILY_SUMMARY_PROTO"
        return try {
            val response = client.request(
                PftpRequest.PbPFtpOperation.newBuilder()
                    .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                    .setPath(dailySummaryFilePath)
                    .build()
                    .toByteArray()
            )
            val proto = DailySummary.PbDailySummary.parseFrom(response.toByteArray())
            when (caloriesType) {
                CaloriesType.ACTIVITY -> proto.activityCalories
                CaloriesType.TRAINING -> proto.trainingCalories
                CaloriesType.BMR -> proto.bmrCalories
            }
        } catch (error: Throwable) {
            BleLogger.w(TAG, "readSpecificCaloriesFromDayDirectory() failed for path: $dailySummaryFilePath, error: $error")
            0
        }
    }

    /**
     * Read and return activity samples data for a given date.
     */
    suspend fun readActivitySamplesDataFromDayDirectory(
        client: BlePsFtpClient,
        date: LocalDate
    ): PolarActivitySamplesDayData {
        BleLogger.d(TAG, "readActivitySamplesDataFromDayDirectory: $date")
        val activityFileDir = "$ARABICA_USER_ROOT_FOLDER${date.format(dateFormatter)}/$ACTIVITY_DIRECTORY"
        val dayData = PolarActivitySamplesDayData()
        val sampleList = mutableListOf<PolarActivitySamplesData>()

        return try {
            val files = listFiles(client, activityFileDir) { entry ->
                entry.matches(Regex("^$activityFileDir/")) ||
                        entry == "ASAMPL" ||
                        entry.contains(".BPB")
            }.toList()

            if (files.isEmpty()) {
                BleLogger.w(TAG, "readActivitySamplesDataFromDayDirectory() could not find files to read for date $date.")
                return dayData
            }

            for (file in files) {
                try {
                    val response = client.request(
                        PftpRequest.PbPFtpOperation.newBuilder()
                            .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                            .setPath(file)
                            .build()
                            .toByteArray()
                    )
                    val proto = ActivitySamples.PbActivitySamples.parseFrom(response.toByteArray())
                    val activitySamplesData = PolarActivitySamplesData().apply {
                        startTime = PolarTimeUtils.pbLocalDateTimeToLocalDateTime(proto.startTime)
                        stepSamples = proto.stepsSamplesList
                        stepRecordingInterval = PolarTimeUtils.pbDurationToInt(proto.stepsRecordingInterval) / 1000
                        metSamples = proto.metSamplesList
                        metRecordingInterval = PolarTimeUtils.pbDurationToInt(proto.metRecordingInterval) / 1000
                        activityInfoList = parsePbActivityInfo(proto.activityInfoList)
                    }
                    sampleList.add(activitySamplesData)
                } catch (error: Throwable) {
                    BleLogger.w(TAG, "readActivitySamplesDataFromDayDirectory() failed for file: $file, error: $error")
                    return dayData
                }
            }

            dayData.polarActivitySamplesDataList = sampleList
            dayData
        } catch (error: Throwable) {
            BleLogger.w(TAG, "readActivitySamplesDataFromDayDirectory() failed while listing files, error occurred $error.")
            dayData
        }
    }

    /**
     * Read and return daily summary data for a given date.
     */
    suspend fun readDailySummaryDataFromDayDirectory(
        client: BlePsFtpClient,
        date: LocalDate
    ): PolarDailySummaryData {
        BleLogger.d(TAG, "readDailySummaryDataFromDayDirectory: $date")
        val dailySummaryFilePath =
            "$ARABICA_USER_ROOT_FOLDER${date.format(dateFormatter)}/$DAILY_SUMMARY_DIRECTORY$DAILY_SUMMARY_PROTO"
        return try {
            val response = client.request(
                PftpRequest.PbPFtpOperation.newBuilder()
                    .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                    .setPath(dailySummaryFilePath)
                    .build()
                    .toByteArray()
            )
            val proto = DailySummary.PbDailySummary.parseFrom(response.toByteArray())
            parsePbDailySummary(proto)
        } catch (error: Throwable) {
            BleLogger.w(TAG, "readDailySummaryDataFromDayDirectory() failed for file: $dailySummaryFilePath, error: $error")
            PolarDailySummaryData()
        }
    }

    private fun listFiles(
        client: BlePsFtpClient,
        folderPath: String = "/",
        condition: PolarFileUtils.FetchRecursiveCondition
    ): Flow<String> {
        var path = folderPath
        if (path.firstOrNull() != '/') path = "/$path"
        if (path.lastOrNull() != '/') path = "$path/"

        return fetchRecursively(client, path, condition)
            .map { it.first }
            .catch { throwable ->
                BleLogger.w(TAG, "listFiles failed for $path: $throwable")
                emitAll(emptyFlow())
            }
    }

    private fun fetchRecursively(
        client: BlePsFtpClient,
        path: String,
        condition: PolarFileUtils.FetchRecursiveCondition
    ): Flow<Pair<String, Long>> = flow {
        val byteArrayOutputStream: ByteArrayOutputStream = client.request(
            PftpRequest.PbPFtpOperation.newBuilder()
                .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                .setPath(path)
                .build()
                .toByteArray()
        )
        val dir = PbPFtpDirectory.parseFrom(byteArrayOutputStream.toByteArray())

        for (entry in dir.entriesList) {
            if (condition.include(entry.name)) {
                val entryPath = path + entry.name
                if (entryPath.endsWith("/")) {
                    emitAll(fetchRecursively(client, entryPath, condition))
                } else {
                    emit(entryPath to entry.size)
                }
            }
        }
    }
}

File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/PolarActivityUtils.swift
//  Copyright © 2024 Polar. All rights reserved.

import Foundation

private let ARABICA_USER_ROOT_FOLDER = "/U/0/"
private let ACTIVITY_DIRECTORY = "ACT/"
private let DAILY_SUMMARY_DIRECTORY = "DSUM/"
private let DAILY_SUMMARY_PROTO = "DSUM.BPB"
private let dateFormat: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyyMMdd"
    formatter.locale = Locale(identifier: "en_US_POSIX")
    return formatter
}()
private let TAG = "PolarActivityUtils"

internal class PolarActivityUtils {

    /// Read step count for given date.
    static func readStepsFromDayDirectory(client: BlePsFtpClient, date: Date) async throws -> Int {
        BleLogger.trace(TAG, "readStepsFromDayDirectory: \(date)")
        let activityFileDir = "\(ARABICA_USER_ROOT_FOLDER)\(dateFormat.string(from: date))/\(ACTIVITY_DIRECTORY)"
        let filePaths: [String]
        do {
            filePaths = try await listFiles(client: client, folderPath: activityFileDir) { entry in
                entry.matches("^\(activityFileDir)") || entry == "ASAMPL" || entry.contains(".BPB")
            }
        } catch {
            BleLogger.error("Failed to list Activity sample files.")
            return 0
        }
        guard !filePaths.isEmpty else { return 0 }
        var stepCount: UInt32 = 0
        for path in filePaths {
            let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = path }
            do {
                let response = try await client.request(try operation.serializedBytes())
                let proto = try Data_PbActivitySamples(serializedBytes: Data(response))
                stepCount += proto.stepsSamples.reduce(0, +)
            } catch {
                BleLogger.error("readStepsFromDayDirectory() failed for path: \(activityFileDir), error: \(error)")
                return 0
            }
        }
        return Int(stepCount)
    }

    /// Read distance in meters for given date.
    static func readDistanceFromDayDirectory(client: BlePsFtpClient, date: Date) async throws -> Float {
        BleLogger.trace(TAG, "readDistanceFromDayDirectory: \(date)")
        let dailySummaryFilePath = "\(ARABICA_USER_ROOT_FOLDER)\(dateFormat.string(from: date))/\(DAILY_SUMMARY_DIRECTORY)\(DAILY_SUMMARY_PROTO)"
        let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = dailySummaryFilePath }
        do {
            let response = try await client.request(try operation.serializedBytes())
            let proto = try Data_PbDailySummary(serializedBytes: Data(response))
            return Float(proto.activityDistance)
        } catch {
            BleLogger.error("readDistanceFromDayDirectory() failed for path: \(dailySummaryFilePath), error: \(error)")
            return 0
        }
    }

    /// Read active time for given date.
    static func readActiveTimeFromDayDirectory(client: BlePsFtpClient, date: Date) async throws -> PolarActiveTimeData {
        BleLogger.trace(TAG, "readActiveTimeFromDayDirectory: \(date)")
        let dailySummaryFilePath = "\(ARABICA_USER_ROOT_FOLDER)\(dateFormat.string(from: date))/\(DAILY_SUMMARY_DIRECTORY)\(DAILY_SUMMARY_PROTO)"
        let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = dailySummaryFilePath }
        do {
            let response = try await client.request(try operation.serializedBytes())
            let proto = try Data_PbDailySummary(serializedBytes: Data(response))
            return PolarActiveTimeData(
                date: date,
                timeNonWear: PolarActiveTime.fromProto(proto.activityClassTimes.timeNonWear),
                timeSleep: PolarActiveTime.fromProto(proto.activityClassTimes.timeSleep),
                timeSedentary: PolarActiveTime.fromProto(proto.activityClassTimes.timeSedentary),
                timeLightActivity: PolarActiveTime.fromProto(proto.activityClassTimes.timeLightActivity),
                timeContinuousModerateActivity: PolarActiveTime.fromProto(proto.activityClassTimes.timeContinuousModerate),
                timeIntermittentModerateActivity: PolarActiveTime.fromProto(proto.activityClassTimes.timeIntermittentModerate),
                timeContinuousVigorousActivity: PolarActiveTime.fromProto(proto.activityClassTimes.timeContinuousVigorous),
                timeIntermittentVigorousActivity: PolarActiveTime.fromProto(proto.activityClassTimes.timeIntermittentVigorous)
            )
        } catch {
            BleLogger.error("readActiveTimeFromDayDirectory() failed for path: \(dailySummaryFilePath), error: \(error)")
            return PolarActiveTimeData(date: date, timeNonWear: PolarActiveTime())
        }
    }

    /// Read calories for given date.
    static func readCaloriesFromDayDirectory(client: BlePsFtpClient, date: Date, caloriesType: CaloriesType) async throws -> Int {
        BleLogger.trace(TAG, "readCaloriesFromDayDirectory: \(date), type: \(caloriesType)")
        let dailySummaryFilePath = "\(ARABICA_USER_ROOT_FOLDER)\(dateFormat.string(from: date))/\(DAILY_SUMMARY_DIRECTORY)\(DAILY_SUMMARY_PROTO)"
        let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = dailySummaryFilePath }
        do {
            let response = try await client.request(try operation.serializedBytes())
            let proto = try Data_PbDailySummary(serializedBytes: Data(response))
            switch caloriesType {
            case .activity: return Int(proto.activityCalories)
            case .training: return Int(proto.trainingCalories)
            case .bmr:      return Int(proto.bmrCalories)
            }
        } catch {
            BleLogger.error("readCaloriesFromDayDirectory() failed for path: \(dailySummaryFilePath), error: \(error)")
            return 0
        }
    }

    /// Read and return activity samples data for a given date.
    static func readActivitySamplesDataFromDayDirectory(client: BlePsFtpClient, date: Date) async throws -> PolarActivityDayData {
        BleLogger.trace(TAG, "readActivitySamplesDataFromDayDirectory: \(date)")
        let activityFileDir = "\(ARABICA_USER_ROOT_FOLDER)\(dateFormat.string(from: date))/\(ACTIVITY_DIRECTORY)"
        let filePaths: [String]
        do {
            filePaths = try await listFiles(client: client, folderPath: activityFileDir) { entry in
                entry.matches("^\(activityFileDir)") || entry == "ASAMPL" || entry.contains(".BPB")
            }
        } catch {
            if error.localizedDescription.contains("103") {
                BleLogger.error("No activity files found for date: \(dateFormat.string(from: date))")
                return PolarActivityDayData(polarActivityDataList: [])
            }
            BleLogger.error("Failed to list activity sample files.")
            throw error
        }
        guard !filePaths.isEmpty else {
            return PolarActivityDayData(polarActivityDataList: [])
        }
        var polarActivityDataList: [PolarActivityData] = []
        for path in filePaths {
            let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = path }
            let response = try await client.request(try operation.serializedBytes())
            let proto = try Data_PbActivitySamples(serializedBytes: Data(response))
            let activitySamplesData = PolarActivityData.PolarActivitySamples(
                startTime: try PolarTimeUtils.pbLocalDateTimeToDate(pbLocalDateTime: proto.startTime),
                metRecordingInterval: PolarTimeUtils.pbDurationToMillis(pbDuration: proto.metRecordingInterval)/1000,
                metSamples: proto.metSamples,
                stepRecordingInterval: PolarTimeUtils.pbDurationToMillis(pbDuration: proto.stepsRecordingInterval)/1000,
                stepSamples: proto.stepsSamples,
                activityInfoList: try PolarActivityData.parsePbActivityInfoList(activityInfoList: proto.activityInfo)
            )
            let activityData = PolarActivityData()
            activityData.samples = activitySamplesData
            polarActivityDataList.append(activityData)
        }
        return PolarActivityDayData(polarActivityDataList: polarActivityDataList)
    }

    /// Read daily summary data for given date. Returns nil if not found (error 103).
    static func readDailySummaryDataFromDayDirectory(client: BlePsFtpClient, date: Date) async throws -> PolarDailySummary? {
        BleLogger.trace(TAG, "readDailySummaryDataFromDayDirectory: \(date)")
        let dailySummaryFilePath = "\(ARABICA_USER_ROOT_FOLDER)\(dateFormat.string(from: date))/\(DAILY_SUMMARY_DIRECTORY)\(DAILY_SUMMARY_PROTO)"
        let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = dailySummaryFilePath }
        do {
            let response = try await client.request(try operation.serializedBytes())
            let proto = try Data_PbDailySummary(serializedBytes: Data(response))
            return try PolarDailySummary.fromProto(proto: proto)
        } catch {
            if error.localizedDescription.contains("103") {
                BleLogger.error("No activity files found for date: \(dateFormat.string(from: date))")
                return nil
            }
            BleLogger.error("Read daily summary failed for date: \(date), error: \(error)")
            throw PolarErrors.fileError(description: error.localizedDescription)
        }
    }

    // MARK: - Private helpers

    private static func listFiles(
        client: BlePsFtpClient,
        folderPath: String = "/",
        condition: @escaping (_ p: String) -> Bool
    ) async throws -> [String] {
        var path = folderPath
        if path.first != "/" { path.insert("/", at: path.startIndex) }
        if path.last  != "/" { path.insert("/", at: path.endIndex) }
        let entries = try await fetchRecursive(path, client: client, condition: condition)
        return entries.map { $0.name }
    }

    private static func fetchRecursive(
        _ path: String,
        client: BlePsFtpClient,
        condition: @escaping (_ p: String) -> Bool
    ) async throws -> [(name: String, size: UInt64)] {
        var operation = Protocol_PbPFtpOperation()
        operation.command = .get
        operation.path = path
        let request = try operation.serializedData()
        let data = try await client.request(request)
        let dir = try Protocol_PbPFtpDirectory(serializedBytes: data as Data)
        var results: [(name: String, size: UInt64)] = []
        for entry in dir.entries where condition(entry.name) {
            let fullPath = path + entry.name
            if fullPath.hasSuffix("/") {
                let children = try await fetchRecursive(fullPath, client: client, condition: condition)
                results.append(contentsOf: children)
            } else {
                results.append((name: fullPath, size: entry.size))
            }
        }
        return results
    }
}
